h2o-algos/src/main/java/hex/tree/TreeHandler.java

```diff
@@ -0,0 +1,161 @@
+        nodesToTraverse.add(sharedTreeSubgraph.rootNode);
+        append(stf.rightChildren, stf.leftChildren,
+                stf.descriptions, nodesToTraverse, -1, false);
+
+        return stf;
+    }
+
+    private static void append(final int[] rightChildren, final int[] leftChildren, final String[] nodesDescriptions,
+                               List<SharedTreeNode> nodesToTraverse, int pointer, boolean visitedRoot) {
+        if(nodesToTraverse.isEmpty()) return;
+
+        List<SharedTreeNode> discoveredNodes = new ArrayList<>();
+
+        for (SharedTreeNode node : nodesToTraverse) {
+            pointer++;
+            final SharedTreeNode leftChild = node.getLeftChild();
+            final SharedTreeNode rightChild = node.getRightChild();
+            if(visitedRoot){
+                nodesDescriptions[pointer] = serializeNodeDescription(node);
+            } else {
+                nodesDescriptions[pointer] = "Root node";
+                visitedRoot = true;
+            }
+
+            if (leftChild != null) {
+                discoveredNodes.add(leftChild);
+                leftChildren[pointer] = leftChild.getNodeNumber();
+            } else {
+                leftChildren[pointer] = NO_CHILD;
+            }
+
+            if (rightChild != null) {
+                discoveredNodes.add(rightChild);
+                rightChildren[pointer] = rightChild.getNodeNumber();
+            } else {
+                rightChildren[pointer] = NO_CHILD;
+            }
+        }
+
+        append(rightChildren, leftChildren, nodesDescriptions, discoveredNodes, pointer, true);
+    }
+
+    private static String serializeNodeDescription(final SharedTreeNode node) {
+        final StringBuffer nodeDescriptionBuffer = new StringBuffer();
+
+        if (!Float.isNaN(node.getParent().getSplitValue())) {
+            nodeDescriptionBuffer.append(node.getParent().getColName());
+            if (node.isLeftward()) {
+                nodeDescriptionBuffer.append(" < ");
+            } else {
+                nodeDescriptionBuffer.append(" >= ");
+            }
+            nodeDescriptionBuffer.append(node.getParent().getSplitValue());
+            return nodeDescriptionBuffer.toString();
+        }
+
+        final BitSet childInclusiveLevels = node.getInclusiveLevels();
+        final int cardinality = childInclusiveLevels.cardinality();
+        if ((cardinality > 0)) {
+            int bitsignCounter = 0;
+            for (int i = childInclusiveLevels.nextSetBit(0); i >= 0; i = childInclusiveLevels.nextSetBit(i + 1)) {
+                nodeDescriptionBuffer.append(node.getParent().getDomainValues()[i]);
+                if (bitsignCounter != cardinality - 1) nodeDescriptionBuffer.append(", ");
+                bitsignCounter++;
+            }
+        }
+
+        return nodeDescriptionBuffer.toString();
+
+    }
+
+    public static class TreeProperties {
+
+        private int[] leftChildren;
```

Comment:
Please use underscore notation for naming fields in pojos (eg. _descriptions).

Reply:
Done.